### PR TITLE
use vectorized loads and stores for all datatypes in torch.cat

### PR DESCRIPTION
**Pull request by ngimel**
Enable vectorized stores in cat whenever possible.
Unforunately, cat on the last dim still struggles to reach peak bw, when last dim sizes are small, so writes from the different threads are not coalesced. Still, it's about 15% gain for the shapes that are supported and where just vectorized reads weren't enough (where the catted slices are multiple of 16-byte alignment), dim0 cats remain approximately the same
The kernel is pretty much copy-paste of CatArrayBatchedCopy_contig kernel, with regular loads/stores replaced by vectorized loads/stores, and necessary adjustments done to the offset calculation to pretend that tensor consists of alignment-sized elements. 

TODO: additional testing, the test failure where some intermediate tensor was of size [1,1] with strides [1024, 1024] and thus took vectorized path was pretty unexpected

### CODE REVIEW

File: aten/src/ATen/native/cuda/Shape.cu [hunk 1]
```diff
@@ -226,6 +226,39 @@ __global__ void CatArrayBatchedCopy_contig(
...
+
+template <typename T, typename IndexType, int Dims, int batch_size, int stride_size, int alignment, int elems_per_vec>
+__global__ void CatArrayBatchedCopy_vectorized(
+    char* output,
+    CatArrInputTensorMetadata<T, IndexType, batch_size, stride_size> inputs,
+    TensorSizeStride<IndexType, CAT_ARRAY_MAX_INPUT_DIMS> os,
+    const int concatDim,
+    IndexType trailingSize) {
+
+    IndexType tid = blockIdx.x * blockDim.x + threadIdx.x;
```
Comment (eqy) [inline aten/src/ATen/native/cuda/Shape.cu:238]: maybe needs a cast to `IndexType`? (our old friend)

**Review by eqy — APPROVED** [review]

File: aten/src/ATen/native/cuda/Shape.cu (hunk 1 above)
Comment (ngimel) [reply aten/src/ATen/native/cuda/Shape.cu:238]: we come here only if all can be done in int32 but I can do it, just to be safe

Comment (ngimel) [reply aten/src/ATen/native/cuda/Shape.cu:238]: on second thought, I'll leave it like this for now, because it's in all the kernels, so we should change that in a separate PR

File: aten/src/ATen/native/cuda/Shape.cu [hunk 2]
```diff
@@ -296,12 +328,27 @@ void parallel_cat(const Tensor &out, const MaterializedITensorListRef& inputs, i
   scalar_t *data = (scalar_t *)(out.mutable_data_ptr());
   CatArrInputTensorMetadata<scalar_t, unsigned int, batch_size, stride_size> catMetaData;
   TensorSizeStride<unsigned int, CAT_ARRAY_MAX_INPUT_DIMS> outputParam;
+  // If all batches are contiguous we can call a specialized implementation
+  // which requires the input tensor addresses to be aligned to a
+  // 16 Byte boundary.
+
+  const bool isContig = stride_size == 1;
```
Comment (Skylion007) [inline aten/src/ATen/native/cuda/Shape.cu:335]: stride_size is a template param, shouldn't this also be constexpr?

File: aten/src/ATen/native/cuda/Shape.cu [hunk 3]
```diff
@@ -296,12 +328,27 @@ void parallel_cat(const Tensor &out, const MaterializedITensorListRef& inputs, i
...
   // Next, let's initialize the size, stride arrays for the output Tensor.
+  // for contig case, we'll canonicalize output strides, so that
+  // we don't have arbitrary strides for dims of size 0
+  size_t stride0 = 1;
   if (memory_format == c10::MemoryFormat::Contiguous) {
-    for (int i = 0; i < nDims; ++i) {
+    for (int i = nDims - 1; i >= 0; --i) {
       outputParam.tensorSize[i] = out.size(i);
-      outputParam.tensorStride[i] = out.stride(i);
+      if (isContig) {
```
Comment (Skylion007) [inline aten/src/ATen/native/cuda/Shape.cu:346]: This can then be made an if constexpr too

File: aten/src/ATen/native/cuda/Shape.cu [hunk 4]
```diff
@@ -351,6 +412,7 @@ void parallel_cat(const Tensor &out, const MaterializedITensorListRef& inputs, i
 #ifdef USE_ROCM
       // On ROCm, CatArrayBatchedCopy_contig is faster
       isAligned = false;
+      isInOutAligned = false;
```
[the remark this answers is not in the document]
Comment (jeffdaily) [reply aten/src/ATen/native/cuda/Shape.cu:414]: do you have a benchmark in mind?  operator_benchmark for pt.cat?

[the remark this answers is not in the document]
Comment (ngimel) [reply aten/src/ATen/native/cuda/Shape.cu:414]: I don't know, what did you use last time when you opted out of aligned kernel?

[the remark this answers is not in the document]
Comment (ngimel) [reply aten/src/ATen/native/cuda/Shape.cu:414]: The shapes I see benefit (compared to aligned kernel) are catting along last dim with smallish last dim, e.g. 
```
dtype =torch.bfloat16
device="cuda"

t1 = torch.randn(1024, 1024, 32, device=device, dtype=dtype)
t2 = torch.randn(1024, 1024, 16, device=device, dtype=dtype)
torch.cat([t1, t2], dim=-1)
```
Also narrow dtypes (like bfloat16 or 1-byte types) benefit more from vectorized reads and writes, you can take a look at benchmarks in #150233
